File: spark/src/main/scala/geotrellis/spark/tiling/LayoutScheme.scala
package geotrellis.spark.tiling

import geotrellis.raster._
import geotrellis.spark._
import geotrellis.vector._

/** A LayoutScheme is something that provides LayoutLevels based on an integer id or 
  * an extent and cellsize. The resolution of the tiles for the LayoutLevel returned
  * will not necessarily match the CellSize provided, but an appropriately close
  * selection will be made.
  * 
  * It also provides methods for next zoomed out tile layout level.
  */
trait LayoutScheme {
  def levelFor(extent: Extent, cellSize: CellSize): LayoutLevel
  def zoomOut(level: LayoutLevel): LayoutLevel
  def zoomIn(level: LayoutLevel): LayoutLevel
}

case class LayoutLevel(zoom: Int, layout: LayoutDefinition)

object LayoutLevel {
  implicit def fromTuple(tup: (Int, LayoutDefinition)): LayoutLevel = LayoutLevel(tup._1, tup._2)
}

File: geotrellis-python/geotrellis/spark/tiling/LayoutScheme.py

class LayoutScheme(object):
    def levelFor(extent, cellSize):
        pass
    def zoomOut(level):
        pass
    def zoomIn(level):
        pass

class LayoutLevel(object):
    def __init__(self, zoom, layout):
        self.zoom = zoom
        self.layout = layout
    def __eq__(self, other):
        if not isinstance(other, LayoutLevel):
            return False
        return self.zoom == other.zoom and self.layout == other.layout
    def __hash__(self):
        return hash((self.zoom, self.layout))
